src/evaluation/Run.jl

```diff
@@ -171,19 +172,104 @@ end
+			# It was not defined previously, we try searching modules in our own batch
+			if res isa LoadError && res.error isa UndefVarError
+				to_import_from_batch = mapreduce(union, unresolved_topology.codes) do (_, cache)
+				    union(cache.module_usings_imports.imports,
+					  cache.module_usings_imports.usings)
+				end
+				WorkspaceManager.do_reimports(sn, to_import_from_batch)
+				# Last try and we leave
+				return try_macroexpand() # 3.
+			end
+		end
+		res
+	end
+
+	function analyze_macrocell(cell::Cell, current_symstate)
+		if ExpressionExplorer.join_funcname_parts.(current_symstate.macrocalls) ⊆ ExpressionExplorer.can_macroexpand
+			return current_symstate
+		else
+			result = macroexpand_cell(cell)
+			if typeof(result) <: Exception
+				# if expansion failed, we use the "shallow" symbols state
+				# we could also use ExpressionExplorer.maybe_macroexpand
+				current_symstate
+			else # otherwise, we use the expanded expression + the list of macrocalls
+				expanded_symbols_state = ExpressionExplorer.try_compute_symbolreferences(result)
+				union!(expanded_symbols_state.macrocalls, current_symstate.macrocalls)
+				expanded_symbols_state
+			end
+		end
+	end
+
+	# create new node & new codes for macrocalled cells
+	new_nodes = Dict{Cell,ReactiveNode}(
+		cell => analyze_macrocell(cell, current_symstate) |> ReactiveNode
+		for (cell, current_symstate) in unresolved_topology.unresolved_cells)
+	all_nodes = merge(unresolved_topology.nodes, new_nodes)
+
+	NotebookTopology(nodes=all_nodes, codes=unresolved_topology.codes)
+end
+
+function static_macroexpand(topology::NotebookTopology, cell::Cell, old_symstate)
+	new_symstate = ExpressionExplorer.maybe_macroexpand(topology.codes[cell].parsedcode; recursive=true) |>
+		ExpressionExplorer.try_compute_symbolreferences
+	union!(new_symstate.macrocalls, old_symstate.macrocalls)
+
+	ReactiveNode(new_symstate)
+end
+
+"The same as `resolve_topology` but does not require custom code execution, only works with a few `Base` & `PlutoRunner` macros"
+function static_resolve_topology(topology::NotebookTopology)
+	new_nodes = Dict{Cell,ReactiveNode}(cell => static_macroexpand(topology, cell, symstate) for (cell, symstate) in topology.unresolved_cells)
+	all_nodes = merge(topology.nodes, new_nodes)
+
+	NotebookTopology(nodes=all_nodes, codes=topology.codes)
+end
+
 ###
 # CONVENIENCE FUNCTIONS
 ###
 
+
 "Do all the things!"
 function update_save_run!(session::ServerSession, notebook::Notebook, cells::Array{Cell,1}; save::Bool=true, run_async::Bool=false, prerender_text::Bool=false, kwargs...)
 	old = notebook.topology
-	new = notebook.topology = updated_topology(old, notebook, cells)
 
-	update_dependency_cache!(notebook)
+	old_workspace_name, new_workspace_name = WorkspaceManager.bump_workspace_module((session, notebook))
+
+	unresolved_topology = updated_topology(old, notebook, cells)
+	new = notebook.topology = resolve_topology(session, notebook, unresolved_topology, old_workspace_name)
```

Comment:
To avoid race conditions, we have a `notebook.executetoken` to ensure that all `run_reactive!` calls are consecutive. `bump_workspace_module` and `resolve_topology` should be called _inside_ `run_reactive!`, when the token is taken, to ensure that they happen within the same run.

Right now, it is possible that `update_save_run!` is called while another call is busy, and it will switch workspaces and expand macros during another run.

Reply:
let me take a shot a this...

Reply:
pretty much done now, some cleanup left